modules/core/src/v2/coins/algo.ts

```diff
@@ -390,38 +450,23 @@ export class Algo extends BaseCoin {
   ): Bluebird<SignedTransaction> {
     const self = this;
     return co<SignedTransaction>(function* () {
-      const { txHex, addressVersion, keys, sk, isHalfSigned } = self.verifySignTransactionParams(params);
-      const encodedPublicKeys = _.map(keys, (k) => Address.decode(k).publicKey);
-
-      // decode our unsigned/half-signed tx
-      let transaction;
-      let txToHex;
-      try {
-        txToHex = Buffer.from(txHex, 'base64');
-        const initialDecodedTx = Encoding.decode(txToHex);
-
-        // we need to scrub the txn of sigs for half-signed
-        const decodedTx = isHalfSigned ? initialDecodedTx.txn : initialDecodedTx;
-
-        transaction = Multisig.MultiSigTransaction.from_obj_for_encoding(decodedTx);
-      } catch (e) {
-        throw new Error('transaction needs to be a valid tx encoded as base64 string');
-      }
-
-      // sign our tx
-      let signed = transaction.partialSignTxn({ version: addressVersion, threshold: 2, pks: encodedPublicKeys }, sk);
-
-      // if we have already signed it, we'll have to merge that with our previous tx
-      if (isHalfSigned) {
-        signed = mergeMultisigTransactions([Buffer.from(signed), txToHex]);
+      const { txHex, signers, prv, isHalfSigned, numberSigners } = self.verifySignTransactionParams(params);
+      const factory = accountLib.register(self.getChain(), accountLib.Algo.TransactionBuilderFactory);
+      const txBuilder = factory.from(txHex);
+      txBuilder.numberOfRequiredSigners(numberSigners);
+      txBuilder.sign({ key: prv });
+      txBuilder.setSigners(signers);
+      const transaction: any = yield txBuilder.build();
```

Comment:
nit: Can we avoid the cast to `any` here?

Reply:
same as above.

Reply:
again, I don't think that argument holds water.